src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs

```diff
@@ -0,0 +1,300 @@
+
+                    // Get unprocessed chars from the buffer.
+                    ReadOnlyMemory<char> charsNotProcessedFromPreviousChunk = buffer;
+
+                    int firstIndexToEscape = this.jsonWriter.NeedsEscaping(chunk.Span);
+
+                    // If the buffer is not empty, then we copy the chars from the buffer
+                    // to the current chunk being processed.
+                    if (!charsNotProcessedFromPreviousChunk.IsEmpty)
+                    {
+                        char[] combinedArray = new char[charsNotProcessedFromPreviousChunk.Length + chunk.Length];
+
+                        // Copy chars from charsNotProcessedFromPreviousChunk to the combined array
+                        charsNotProcessedFromPreviousChunk.CopyTo(combinedArray);
+
+                        // Copy chars from chunk to the combined array, starting from the end of charsNotProcessedFromPreviousChunk
+                        chunk.CopyTo(combinedArray.AsMemory().Slice(charsNotProcessedFromPreviousChunk.Length));
+
+                        // Write the chunk depending on the firstIndexToEscape
+                        WriteChunk(combinedArray, isFinalBlock, firstIndexToEscape);
+                    }
+                    else
+                    {
+                        // Write the chunk directly depending on the firstIndexToEscape
+                        WriteChunk(chunk.Span, isFinalBlock, firstIndexToEscape);
+                    }
+
+                    // Flush the writer if the buffer threshold is reached
+                    await this.jsonWriter.FlushIfBufferThresholdReachedAsync().ConfigureAwait(false);
+                }
+            }
+
+            /// <summary>
+            /// Writes a chunk of characters to the JSON writer, either escaped or directly, depending on the presence of characters that need escaping.
+            /// </summary>
+            /// <param name="chunk">The ReadOnlySpan containing the chunk of characters to be written.</param>
+            /// <param name="isFinalBlock">A boolean value indicating whether the current chunk is the final block of characters to be written.</param>
+            /// <param name="firstIndexToEscape">The index of the first character in the chunk that requires escaping, or -1 if no characters need escaping.</param>
+            private void WriteChunk(ReadOnlySpan<char> chunk, bool isFinalBlock, int firstIndexToEscape)
+            {
+                if (firstIndexToEscape != -1)
+                {
+                    this.jsonWriter.WriteEscapedStringChunk(chunk, firstIndexToEscape, isFinalBlock, out int charsNotWrittenFromPreviousChunk);
+
+                    if (charsNotWrittenFromPreviousChunk > 0)
+                    {
+                        // Update the buffer with unprocessed chars from the current chunk.
+                        buffer = chunk.Slice(chunk.Length - charsNotWrittenFromPreviousChunk).ToArray();

```

Comment:
Would a Span<byte> be possible to use as a buffer in place of byte[] so we don't have to do .ToArray()?